### ISSUE TITLE
Multi-Turn Phi35 mini instruct training

### ISSUE THREAD

**Issue by chwunderpers** [huggingface/trl#4922]
Hi all,

I try to train a model with a multi-turn conversation.  Unfortunately i´m facing an issue in which for me it seems, that the model is ignoring the multi-turn conversations at all to lear from.

the outputs of the model are intents like < rag>, < command>, ... or the special case < followup>. (Syntax is slightly different, but the purpose stays the same.

Now I have done similar training without multi-turn and the models performance is awesome, also with these kind of outputs. So this is a proven finetuning.

But as soon as I feed it with followup history related question, it doesn´t learn and from the 5000 test samples, it seems like it is always mirroring the first assistant output, rather than <followup>

here is an example:

```json
{
    "input": {
        "messages": [
            {
                "role": "system",
                "content": "You are an Intent Recognition assistant. Classify each user query into the correct function.\n\nIMPORTANT: If the user's question is a follow-up or clarification about the SAME topic as the previous exchange, use <followup>();. Only use base intents (rag, command, unknown) for NEW topics.\n\nFunctions:\n- <rag>(); - Questions about documentation, features, how-to guides (NEW topic)\n- <command>(xxx=N); - reqeuests for XXXX\n- <followup>(); - Any continuation/clarification of the previous topic\n- <unknown>(); - Off-topic or unrelated questions"
            },
            {
                "role": "user",
                "content": "How do I stop a charging session early?"
            },
            {
                "role": "assistant",
                "content": "<rag>();"
            },
            {
                "role": "user",
                "content": "Can I do it from inside the car?"
            },
            {
                "role": "assistant",
                "content": "<followup>();"
            },
            {
                "role": "user",
                "content": "Does the cable unlock automatically then?"
            },
            {
                "role": "assistant",
                "content": "<followup>();"
            }
        ]
    },
    "response": "<rag>();"
}
```

As you can see from this conversation, the first assistant message is < rag>();, but the model predicted (thats the response field) rag rather than followup. Of course the last assistant message is being removed before inference.

As you can see, any other example (system prompt removed for readability is spot on:
```json
[
    {
        "role": "user",
        "content": "How do I change the language of the infotainment system?"
    },
    {
        "role": "assistant",
        "content": "<rag>();"
    },
    {
        "response": "<rag>();"
    }
]
```

So there is nothing wrong per say with the training from what I can see, it just seems to be something wrong with the histroy.

There are thousands of examples with history, without histroy. With rag or other commands as the expected results with history, of course  more with followup as the expected result. Not all conversations are followups if the user asks a comptely different topic at the end of a conversation and so on.
Training metrics are also very good, all grpahs are gentle and seem to be right.

Is there any trick I have to follow to get the conversation consiedered in the training. 

Using standard sft

        trainer = SFTTrainer(
            model=self.model,
            tokenizer=self.tokenizer,
            train_dataset=formatted_train_dataset,
            eval_dataset=formatted_valid_dataset, 
            peft_config=peft_config,
            args=training_args,
            dataset_text_field="text",
            callbacks=[LogLossMLFlowCallback()] if self.settings.use_mlflow else []
        )
        

The text whats going into the trainer is like which is also ok in my opintion:
<|system|>
You are an Intent Recognition assistant. Classify each user query into the correct function.

IMPORTANT: If the user's question is a follow-up or clarification about the SAME topic as the previous exchange, use <followup>();. Only use base intents (rag, command , unknown) for NEW topics.

Functions:
- <rag>(); - Questions about documentation, features, how-to guides (NEW topic)
- <command>(); - this is redacted ...
- <followup>(); - Any continuation/clarification of the previous topic
- <unknown>(); - Off-topic or unrelated questions<|end|>
<|user|>
Is there mud or dirt on the road ahead?<|end|>
<|assistant|>
<admonitor>();<|end|>
<|user|>
How far does it extend?<|end|>
<|assistant|>
<followup>();<|end|>
<|endoftext|>

**Comment by Saad-Mallebhari** [comment]
@chwunderpers 
The issue is likely that `SFTTrainer` with `dataset_text_field` trains on all tokens in the text, including every intermediate assistant turn , not just the final response. So the model sees `<rag>();` as a valid training target mid-conversation and learns to predict it regardless of what follows.

To fix this, you need to mask all assistant turns except the last one so the model only learns to predict the final response. Use `dataset_text_field` with a custom data collator, or better, switch to passing a `messages` column and use `DataCollatorForCompletionOnlyLM` with the assistant response token as the response template:
```python
from trl import DataCollatorForCompletionOnlyLM

collator = DataCollatorForCompletionOnlyLM(
    response_template="<|assistant|>",
    tokenizer=tokenizer,
)

trainer = SFTTrainer(
    ...
    data_collator=collator,
)
```
This ensures only the final assistant turn contributes to the loss, which should fix the model always mirroring the first assistant output.

**Comment by qgallouedec** [comment]
If you want to train on the last turn only, use a [prompt-completion dataset](https://huggingface.co/docs/trl/en/dataset_formats#prompt-completion) where the completion is the last turn. See the [Train on completion only](https://huggingface.co/docs/trl/en/sft_trainer#train_on_completion_only).

---

If you want to train on assistant turns only, use `assistant_only_loss`

```python
training_args = SFTConfig(assistant_only_loss=True)
```

See the [Train on assistant messages only](https://huggingface.co/docs/trl/en/sft_trainer#train_on_assistant_messages_only).

--

I discourage the approach proposed by @Saad-Mallebhari. It may have been valid several versions ago, but it is now largely outdated:
- `dataset_text_field` is not used for conversational data, so it is not relevant here.
- Applying the chat template before passing the dataset to the SFTTrainer is strongly discouraged. Depending on the chat template, doing so can lead to incorrect handling of special tokens and other subtle formatting issues. You should rely on SFTTrainer to apply the template properly.
- `DataCollatorForCompletionOnlyLM` was removed from TRL some time ago in favor of the approaches mentioned above.

**Comment by qgallouedec** [comment]
Hi @chwunderpers, sorry for the slow follow-up, we're catching up on the backlog.

This isn't a TRL bug: with default SFT on conversational data the loss is computed over the whole sequence, so every intermediate assistant turn (e.g. the first `<rag>();`) is itself a training target. Which is exactly why the model happily emits it mid-conversation.

As noted above, set `SFTConfig(assistant_only_loss=True)` to train on assistant turns only, or use a prompt-completion dataset if you only want the final turn as the target. And let `SFTTrainer` apply the chat template rather than pre-formatting a `text` field!
